File: app/src/main/java/br/com/jonathanzanella/myexpenses/expense/ExpenseAdapterPresenter.java
package br.com.jonathanzanella.myexpenses.expense;

import org.apache.commons.lang3.StringUtils;
import org.joda.time.DateTime;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class ExpenseAdapterPresenter {
	private final ExpenseRepository repository;
	private final ExpenseAdapter adapter;

	private List<Expense> receipts;
	private List<Expense> receiptsFiltered;

	ExpenseAdapterPresenter(ExpenseAdapter adapter, ExpenseRepository repository) {
		this.repository = repository;
		this.adapter = adapter;
	}

	private void loadExpenses(DateTime date) {
		receipts = repository.monthly(date);
		receiptsFiltered = receipts;
	}

	List<Expense> getExpenses(boolean invalidateCache, DateTime date) {
		if(invalidateCache)
			loadExpenses(date);
		if(receiptsFiltered == null)
			return null;
		return Collections.unmodifiableList(receiptsFiltered);
	}

	public void filter(String filter) {
		if(filter == null || filter.compareTo("") == 0) {
			receiptsFiltered = receipts;
			return;
		}

		receiptsFiltered = new ArrayList<>();
		for (Expense bill : receipts) {
			if(StringUtils.containsIgnoreCase(bill.getName(), filter))
				receiptsFiltered.add(bill);
		}
	}
}

File: app/src/main/java/br/com/jonathanzanella/myexpenses/expense/ExpenseAdapterPresenter.kt
package br.com.jonathanzanella.myexpenses.expense

import org.apache.commons.lang3.StringUtils
import org.joda.time.DateTime
import java.util.*

class ExpenseAdapterPresenter(private val repository: ExpenseRepository) {
    private var receipts: MutableList<Expense>? = null
    private var receiptsFiltered: MutableList<Expense>? = null

    private fun loadExpenses(date: DateTime) {
        receipts = repository.monthly(date) as MutableList<Expense>
        receiptsFiltered = receipts
    }

    fun getExpenses(invalidateCache: Boolean, date: DateTime?): List<Expense> {
        if (invalidateCache)
            loadExpenses(date!!)
        return Collections.unmodifiableList(receiptsFiltered ?: ArrayList())
    }

    fun filter(filter: String) {
        if (filter.compareTo("") == 0) {
            receiptsFiltered = receipts
            return
        }

        receiptsFiltered = ArrayList()
        receipts!!
                .filter { StringUtils.containsIgnoreCase(it.name, filter) }
                .forEach { receiptsFiltered!!.add(it) }
    }
}
